File: problems/tree/0104-maximum-depth-of-binary-tree.js
/**
 * #104 Maximum Depth of Binary Tree
 * Recursive contract: return the deepest path length from this node to a leaf.
 * Time: O(n) | Space: O(h) call stack
 *
 * @param {TreeNode|null} root
 * @return {number}
 */
function maxDepth(root) {
  if (root === null) return 0
  return 1 + Math.max(maxDepth(root.left), maxDepth(root.right))
}

module.exports = maxDepth

File: problems/tree/0104-maximum-depth-of-binary-tree.py
"""#104 Maximum Depth of Binary Tree
Return one plus the greater of the two subtree depths.
Time: O(n) | Space: O(h) call stack.
"""


class Solution:
    def maxDepth(self, root):
        if root is None:
            return 0
        return 1 + max(self.maxDepth(root.left), self.maxDepth(root.right))
